src/librustc_typeck/check/mod.rs

```diff
@@ -753,7 +753,12 @@ fn check_mod_item_types(tcx: TyCtxt<'_>, module_def_id: DefId) {
 fn typeck_item_bodies(tcx: TyCtxt<'_>, crate_num: CrateNum) {
     debug_assert!(crate_num == LOCAL_CRATE);
     tcx.par_body_owners(|body_owner_def_id| {
-        tcx.ensure().typeck_tables_of(body_owner_def_id);
+        if let Some(hir::Node::AnonConst(_)) = tcx.hir().get_if_local(body_owner_def_id) {
+            // We don't want to incorrectly suggest constraining type parameters for array lengths.
+            tcx.ensure().const_typeck_tables_of(body_owner_def_id);
+        } else {
+            tcx.ensure().typeck_tables_of(body_owner_def_id);
+        }
```

Comment:
This (and changes below) obligates us to keep these two in sync, and if not, we risk having diagnostics improvements result in language changes. Imo the risk here is not worth it, and we should find a less risky way to do it or forego the improvements.

Reply:
It's also essentially a hack around bugs in the compiler, so we're adding tech debt here to cleanup later.